pyquil/api/_compiler.py

```diff
@@ -129,63 +129,84 @@ def _collect_memory_descriptors(program: Program) -> Dict[str, ParameterSpec]:
+        if not quilc_endpoint.startswith('tcp://'):
             raise ValueError(f"PyQuil versions >= 2.4 can only talk to quilc "
                              f"versions >= 1.4 over network RPCQ.  You've supplied the "
-                             f"endpoint '{endpoint}', but this doesn't look like a network "
+                             f"endpoint '{quilc_endpoint}', but this doesn't look like a network "
                              f"ZeroMQ address, which has the form 'tcp://domain:port'. "
                              f"You might try clearing (or correcting) your COMPILER_URL "
                              f"environment variable and removing (or correcting) the "
                              f"compiler_server_address line from your .forest_config file.")
 
-        self.client = Client(endpoint, timeout=timeout)
+        self.quilc_client = Client(quilc_endpoint, timeout=timeout)
+        if qpu_compiler_endpoint is not None:
+            self.qpu_compiler_client = Client(qpu_compiler_endpoint, timeout=timeout)
+        else:
+            self.qpu_compiler_client = None
+            warnings.warn("It looks like you are initializing a QPUCompiler object without a "
+                          "qpu_compiler_address. If you didn't do this manually, then "
+                          "you probably don't have a qpu_compiler_address entry in your "
+                          "~/.forest_config file, meaning that you are not engaged to the QPU.")
         self.target_device = TargetDevice(isa=device.get_isa().to_dict(),
                                           specs=device.get_specs().to_dict())
         self.name = name
 
     def get_version_info(self) -> dict:
-        return self.client.call('get_version_info')
+        quilc_version_info = self.quilc_client.call('get_version_info')
+        if self.qpu_compiler_client:
+            qpu_compiler_version_info = self.qpu_compiler_client.call('get_version_info')
+            return {'quilc': quilc_version_info, 'qpu_compiler': qpu_compiler_version_info}
+        return {'quilc': quilc_version_info}
 
     @_record_call
     def quil_to_native_quil(self, program: Program) -> Program:
         request = NativeQuilRequest(quil=program.out(), target_device=self.target_device)
-        response = self.client.call('quil_to_native_quil', request).asdict()  # type: Dict
+        response = self.quilc_client.call('quil_to_native_quil', request).asdict()  # type: Dict
         nq_program = parse_program(response['quil'])
         nq_program.native_quil_metadata = response['metadata']
         nq_program.num_shots = program.num_shots
         return nq_program
 
     @_record_call
-    def native_quil_to_executable(self, nq_program: Program) -> BinaryExecutableResponse:
+    def native_quil_to_executable(self, nq_program: Program) -> Optional[BinaryExecutableResponse]:
+        if not self.qpu_compiler_client:
+            warnings.warn("It looks like you're trying to compile to an executable, but "
```

Comment:
This should just be raising an exception

Reply:
Ah yes.